### ISSUE TITLE
[BUG] `error: Alias bash-language-server is still in node-packages.nix` error when building on current nixos-unstable

### ISSUE THREAD

**Issue by Daru-san** [nix-community/nixvim#1699]
| Field        | Description |
|--------------|-------------|
| Plugin       | `bashls`  | 
| Nixpkgs      | `unstable` | 
| Home Manager | `unstable` | 

- [x] I have read the [FAQ](https://nix-community.github.io/nixvim/user-guide/faq.html) and my bug is not listed there.

## Description

Every time I  attempt to build nixvim on the current revision of nixpkgs on nixos-unstable, the error below is thrown.
I assumed that I had to disable bashls in the lsp options in order to fix this but that does not fix it
```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'nixvim'
         whose name attribute is located at /nix/store/qqwr649pc0qprc9lw2fmdsi1km6p7q2h-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'paths' of derivation 'nixvim'
         at /nix/store/fg9vjx7bl5yj0slbwgw1rjghmlvvi83f-source/wrappers/standalone.nix:52:7:
           51|       name = "nixvim";
           52|       paths = [
             |       ^
           53|         config.finalPackage

       … while evaluating the default value of option `plugins.lsp.servers.bashls.cmd`

       … while evaluating the option `plugins.lsp.servers.bashls.package':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Alias bash-language-server is still in node-packages.nix
```

## Minimal, Reproducible Example (MRE)

```nix
# This applies whether bashls is enabled or not
programs.nixvim = {
  plugins.lsp.servers.bashls.enable = true;
}
```

**Comment by MattSturgeon** [comment]
The issue is we are using an old package name for the bash language server. This should only affect you if you are overriding nixvim's nixpkgs lock (i.e. `nixvim.inputs.nixpkgs.follows`).

Until we fix this in See https://github.com/nix-community/nixvim/pull/1696, you should either remove/comment out your "nixvim follows" line, or rollback your flake.lock to an older nixpkgs revision.

**Comment by Amelia-Mowers** [comment]
Just had this and fixed this issue by removing `nixvim.inputs.nixpkgs.follows` but you have to also run `nix flake update`, or similar command to just update that one flake, as well. Which was not immediately obvious to me.

**Comment by MattSturgeon** [comment]
> Just had this and fixed this issue by removing `nixvim.inputs.nixpkgs.follows` but you have to also run `nix flake update`, or similar command to just update that one flake, as well. Which was not immediately obvious to me.

If you want to apply input changes to `flake.lock` _without_ also updating, you can use `nix flake lock`.

Incidentally, if you want to change what an input points to without modifying your `flake.nix` (e.g. to temporarily test a PR or revert to an old rev) you can use `nix flake lock --override-input <input-name> <input-url>`.

**Comment by niksingh710** [comment]
i have only `    nixvim.url = "github:nix-community/nixvim";` in my inputs. it has been generated by `nix flake init -t github:nix-community/nixvim` i don't see any option where it is specified to be `    nixvim.url = "github:nix-community/nixvim";`

still having the same issue.

**Comment by djacu** [comment]
I was still getting the error after removing the "follows" line. I ended up pinning my nixpkgs input to the same rev as the nixpkgs input for nixvim and that worked fine as expected.

**Comment by bearfm** [comment]
I removed the 'follows' line from mine, went to the directory with my NixOS config, and ran `sudo nix flake update`, and then ran `sudo nixos-rebuild --flake . switch --upgrade` and it built with no issues.

**Comment by MattSturgeon** [comment]
> I was still getting the error after removing the "follows" line

Did you run `nix flake lock` after removing the `follows`?

**Comment by MartinEekGerhardsen** [comment]
After removing the `follows` and running both `nix flake update` and `nix flake lock` multiple times, I still get the same build error. Is there any information I can provide to try to find a solution to this, or would it be easiest to wait for #1696?

**Comment by MattSturgeon** [comment]
> After removing the `follows` and running both `nix flake update` and `nix flake lock` multiple times, I still get the same build error.

If `follows` is not having an effect, try overriding your main nixpkgs input (the one that becomes `pkgs` in your modules) to point to a nixpkgs without this issue. We're currently locked to `57d6973abba7ea108bac64ae7629e7431e0199b6`.

The upstream PR has made it to `nixos-unstable-small`, so that channel _may_ also work (hopefully, anyway).

> Is there any information I can provide to try to find a solution to this

I'm curious which installation method you use for nixvim (nixos, hm, darwin, standalone), it may be this determines whether nixvim evaluate using its own nixpkgs input or the `pkgs` of a wider module system.

**Comment by MartinEekGerhardsen** [comment]
> If `follows` is not having an effect, try overriding your main nixpkgs input (the one that becomes `pkgs` in your modules) to point to a nixpkgs without this issue. We're currently locked to `57d6973abba7ea108bac64ae7629e7431e0199b6`.
Using this commit, I no longer get this error, both with and without follows. Thank you!

> I'm curious which installation method you use for nixvim (nixos, hm, darwin, standalone), it may be this determines whether nixvim evaluate using its own nixpkgs input or the `pkgs` of a wider module system.
I'm using a standalone `nixvim` flake on my system running NixOS `x86_64-linux` and just add the output of the standalone flake to my system packages.

**Comment by meck** [comment]
I'm also seeing this where I got a custom nixvim [flake](https://github.com/meck/mvim) that wraps nixvim this has 
```
nixpkgs.follows = "nixvim/nixpkgs";
```

And just running `nix run .#` works. But if i add it to my nixos config which just have `mvim.url = "github:meck/mvim"` (no follow in the input) this error pops up as if it using the general nixos nixpkgs version, nixos-unstable.

**Comment by purepani** [comment]
The update has now made it into nixpkgs. Now https://github.com/nix-community/nixvim/issues/1702 is stopping me from building, but this is a separate issue.

**Comment by MattSturgeon** [comment]
> The update has now made it into nixpkgs. Now #1702 is stopping me from building, but this is a separate issue.

It's not yet in `nixos-unstable`. It's in all the other channels though: https://nixpk.gs/pr-tracker.html?pr=319882

`nixos-unstable` is [currently 3 days behind master](https://status.nixos.org)

**Comment by iSma** [comment]
Just curious, is there a reason why setting `plugins.lsp.servers.bashls.enable = false;` and/or overriding `plugins.lsp.servers.bashls.package` still triggers this issue? I had a quick look at `_mk-lsp.nix` and from my understanding `package` shouldn't get evaluated if `enabled` is false.

**Comment by MattSturgeon** [comment]
> Just curious, is there a reason why setting `plugins.lsp.servers.bashls.enable = false;` and/or overriding `plugins.lsp.servers.bashls.package` still triggers this issue? I had a quick look at `_mk-lsp.nix` and from my understanding `package` shouldn't get evaluated if `enabled` is false.



> Just curious, is there a reason why setting `plugins.lsp.servers.bashls.enable = false;` and/or overriding `plugins.lsp.servers.bashls.package` still triggers this issue? I had a quick look at `_mk-lsp.nix` and from my understanding `package` shouldn't get evaluated if `enabled` is false.

I haven't looked, but NixOS modules often evaluate a lot of branches that might not get evaluated in non-module nix expressions. This is done to ensure more parts of your config are valid, even some parts you aren't using.

For instance, `lib.mkIf` isn't actually using `if then else`, instead it just returns an attrset with a condition attribute and a config attribute. The module system itself then evaluates the config, regardless of condition being successful. It just only gets merged into the final `config` if the condition was successful.

**Comment by dlip** [comment]
I suspect this might be due to `home-manager.useGlobalPkgs = true;` but is breaks too many things for me to confirm this

**Comment by GaetanLepage** [comment]
Fixed in https://github.com/nix-community/nixvim/pull/1719

**Comment by dlip** [comment]
Since I thought it might be of use to others, I fixed the nixpkg dependency overriding problem by removing the home manger module and just used nixvim as a standalone overlay. Getting extra plugins working was a little tricky:

I add it to my overlay and add my extraPluginsSrc vim plugin inputs to extraSpecialArgs. It uses a trick to get all the inputs starting with "vimplugin-", but you can just pass in an attrset manually if you like.

https://github.com/dlip/nixconfig/blob/e19874c0f1326390708f7003af4c2e07ecf3e45d/overlays/default.nix#L43-L48

I then build them using the nixpkgs in nixvim

https://github.com/dlip/nixconfig/blob/e19874c0f1326390708f7003af4c2e07ecf3e45d/overlays/nixvim/default.nix#L6-L12

And pass in to extraPlugins

https://github.com/dlip/nixconfig/blob/e19874c0f1326390708f7003af4c2e07ecf3e45d/overlays/nixvim/default.nix#L886-L890

then you can just use

```
home.packages = [ nixvim ];
```

**Comment by Toshiro00** [comment]
> The issue is we are using an old package name for the bash language server.
> 
> This should only affect you if you are overriding nixvim's nixpkgs lock (i.e. `nixvim.inputs.nixpkgs.follows`).
> 
> Until we fix this in See #1696, you should either remove/comment out your "nixvim follows" line, or rollback your flake.lock to an older nixpkgs revision.
> 
> EDIT: upstream PR is [NixOS/nixpkgs#319882](https://github.com/NixOS/nixpkgs/pull/319882)

I had same issue with "error: Alias @astrojs/language-server is still in node-packages.nix" just removed "nixpkgs.follows" solved the issue.

**Comment by GaetanLepage** [comment]
> I had same issue with "error: Alias @astrojs/language-server is still in node-packages.nix" just removed "nixpkgs.follows" solved the issue.

Addressed in https://github.com/nix-community/nixvim/pull/2002/commits/8ff21b7e61b5c689e56f58425897209c5d57790c

**Comment by landure** [comment]
For future reference, since it's not documented here. I'm running a system without `nixpkgs.follows` nor flakes. 
I'm a very new nix user.

The issue comes from `nixpkgs` node-packages. 

At the top of `/nix/store/6zw2bds8vf6k74w3mk4j7hxmx4nvd3aj-nixpkgs-24.11/nixpkgs/pkgs/development/node-packages/aliases.nix`, there's a mention to `remove-attr.py`

```nix
### Use `./remove-attr.py [attrname]` in this directory to remove your alias
### from the `nodePackages` set without regenerating the entire file.
```

Running this command fix the issue:

```bash
sudo python3 /nix/store/6zw2bds8vf6k74w3mk4j7hxmx4nvd3aj-nixpkgs-24.11/nixpkgs/pkgs/development/node-packages/remove-attr.py 'bash-language-server'
```

**Comment by MattSturgeon** [comment]
> Running this command fix the issue:
> 
> ```shell
> sudo python3 /nix/store/6zw2bds8vf6k74w3mk4j7hxmx4nvd3aj-nixpkgs-24.11/nixpkgs/pkgs/development/node-packages/remove-attr.py 'bash-language-server'
> ```

I'm very surprised that did anything to your local system. That is something that is supposed to be done in upstream nixpkgs.

I'm also a little confused. This issue was supposed to have been fixed upstream, so I don't know how you ran into it. Is your flake.lock and/or channels fully up to date?

**Comment by landure** [comment]
Hi, keep in mind I'm very new to Nix.

I've written my whole fix process here: [Nix: fix Alias <package> is still in node-packages.nix @ GitHub](https://github.com/biapy/howto-biapy-com/blob/minimalist/content/blog/nix-node-packages-alias.md).

I've encountered the issue by : 

1. Installing nix (unstable). 
2. Launching `lunarvim` with `nix-shell -p 'lunarvim'`
3. Setting user channel to 24.11 stable (and updating channels)
4. Launching `lunarvim` with `nix-shell` -> issue with `neovim` alias.
5. using `--show-trace` to find that the issue takes places in `<nixpkgs>/pkgs/development/node-packages/aliases.nix`
6. reading `aliases.nix`
7. applying the fix I've described, as documented in `aliases.nix`.

I'm not sure this will reproduce the issue, but I believe it's the steps I've taken to encounter it.
